# NLTK Sentiment Analysis

In [5]:
import nltk
nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    |   Package abc is already up-to-date!
[nltk_data]    | Downloading package alpino to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    |   Package alpino is already up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger is already up-
[nltk_data]    |       to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger_ru is already
[nltk_data]    |       up-to-date!
[nltk_data]    | Downloading package basque_grammars to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    | 

[nltk_data]    |   Package movie_reviews is already up-to-date!
[nltk_data]    | Downloading package mte_teip5 to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    |   Package mte_teip5 is already up-to-date!
[nltk_data]    | Downloading package mwa_ppdb to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    |   Package mwa_ppdb is already up-to-date!
[nltk_data]    | Downloading package names to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    |   Package names is already up-to-date!
[nltk_data]    | Downloading package nombank.1.0 to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    |   Package nombank.1.0 is already up-to-date!
[nltk_data]    | Downloading package nonbreaking_prefixes to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    |   Package nonbreaking_prefixes is already up-to-date!
[nltk_data]    | Downloading package nps_chat

[nltk_data]    |   Unzipping corpora\twitter_samples.zip.
[nltk_data]    | Downloading package udhr to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping corpora\udhr.zip.
[nltk_data]    | Downloading package udhr2 to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping corpora\udhr2.zip.
[nltk_data]    | Downloading package unicode_samples to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping corpora\unicode_samples.zip.
[nltk_data]    | Downloading package universal_tagset to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping taggers\universal_tagset.zip.
[nltk_data]    | Downloading package universal_treebanks_v20 to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    | Downloading package vader_lexicon to
[nltk_data]    |     C:\Users\alpid\AppData\Roaming\nltk_data...
[nltk_data]    | D

True

Tokenization  
Tokenization is a text preprocessing step in sentiment analysis that involves breaking down the text into individual words or tokens. This is an essential step in analyzing text data as it helps to separate individual words from the raw text, making it easier to analyze and understand. Tokenization is typically performed using NLTK's built-in `word_tokenize` function, which can split the text into individual words and punctuation marks.

Stop words  
Stop word removal is a crucial text preprocessing step in sentiment analysis that involves removing common and irrelevant words that are unlikely to convey much sentiment. Stop words are words that are very common in a language and do not carry much meaning, such as "and," "the," "of," and "it." These words can cause noise and skew the analysis if they are not removed.

By removing stop words, the remaining words in the text are more likely to indicate the sentiment being expressed. 

Stemming and Lemmatization  
Stemming and lemmatization are techniques used to reduce words to their root forms. Stemming involves removing the suffixes from words, such as "ing" or "ed," to reduce them to their base form. For example, the word "jumping" would be stemmed to "jump." 

Lemmatization, however, involves reducing words to their base form based on their part of speech. For example, the word "jumped" would be lemmatized to "jump," but the word "jumping" would be lemmatized to "jumping" since it is a present participle.

Bag of Words (BoW) Model  
The bag of words model is a technique used in natural language processing (NLP) to represent text data as a set of numerical features. In this model, each document or piece of text is represented as a "bag" of words, with each word in the text represented by a separate feature or dimension in the resulting vector. The value of each feature is determined by the number of times the corresponding word appears in the text.

## 1. Import libraries

In [10]:
# import libraries
import pandas as pd

import nltk

from nltk.sentiment.vader import SentimentIntensityAnalyzer

from nltk.corpus import stopwords

from nltk.tokenize import word_tokenize

from nltk.stem import WordNetLemmatizer


In [11]:
# Load the amazon review dataset

df = pd.read_csv('https://raw.githubusercontent.com/pycaret/pycaret/master/datasets/amazon.csv')

df

reviewText  Positive
0      This is a one of the best apps acording to a b...         1
1      This is a pretty good version of the game for ...         1
2      this is a really cool game. there are a bunch ...         1
3      This is a silly game and can be frustrating, b...         1
4      This is a terrific game on any pad. Hrs of fun...         1
...                                                  ...       ...
19995  this app is fricken stupid.it froze on the kin...         0
19996  Please add me!!!!! I need neighbors! Ginger101...         1
19997  love it!  this game. is awesome. wish it had m...         1
19998  I love love love this app on my side of fashio...         1
19999  This game is a rip off. Here is a list of thin...         0

[20000 rows x 2 columns]

## 2. Preprocess text

In [12]:
# create preprocess_text function
def preprocess_text(text):

    # Tokenize the text.
    # The function first tokenizes the text using the word_tokenize function from the nltk library,
    # which splits the text into individual words.

    tokens = word_tokenize(text.lower()) # text.lower convierte todo el texto en moniuscilas apra no difierecniar


    # Remove stop words. using a list comprehension that checks if each token is not in the list
    # of English stop words provided by the stopwords module in the nltk library.

    filtered_tokens = [token for token in tokens if token not in stopwords.words('english')]


    # Lemmatize the tokens. lemmatizes the remaining tokens using the WordNetLemmatizer class
    # from the nltk library, which reduces each word to its base form.

    lemmatizer = WordNetLemmatizer()

    lemmatized_tokens = [lemmatizer.lemmatize(token) for token in filtered_tokens]


    # Join the tokens back into a string

    processed_text = ' '.join(lemmatized_tokens)

    return processed_text

# apply the function df. The code then applies this preprocess_text function to the reviewText
# column of a pandas DataFrame called df using the apply method, which applies the function
#to each element of the column.
# The resulting processed text is then stored back in the reviewText column of the DataFrame.

df['reviewText'] = df['reviewText'].apply(preprocess_text)
df

reviewText  Positive
0      one best apps acording bunch people agree bomb...         1
1      pretty good version game free . lot different ...         1
2      really cool game . bunch level find golden egg...         1
3      silly game frustrating , lot fun definitely re...         1
4      terrific game pad . hr fun . grandkids love . ...         1
...                                                  ...       ...
19995  app fricken stupid.it froze kindle wont allow ...         0
19996  please add ! ! ! ! ! need neighbor ! ginger101...         1
19997  love ! game . awesome . wish free stuff house ...         1
19998  love love love app side fashion story fight wo...         1
19999  game rip . list thing make better & bull ; fir...         0

[20000 rows x 2 columns]

Notice the changes in the "review text" column as a result of the preprocess_text function that we applied in the above step.

## 3. NLTK Sentiment Analyzer

In [13]:
# initialize NLTK sentiment analyzer from nltk.sentiment.vader

analyzer = SentimentIntensityAnalyzer()

# create get_sentiment function that takes a text string as its input.
# The function calls the polarity_scores method of the analyzer object to obtain a dictionary
# of sentiment scores for the text, which includes a score for positive, negative, and neutral
# sentiment. 

def get_sentiment(text):

    scores = analyzer.polarity_scores(text)

    sentiment = 1 if scores['pos'] > 0 else 0 #The function will then check whether the positive
                                              # score is greater than 0 and returns a sentiment
                                              # score of 1 if it is, and a 0 otherwise.

    return sentiment


# apply get_sentiment function to the reviewText column of the df DataFrame using the apply method.
# This creates a new column called sentiment in the DataFrame, which stores the sentiment score
# for each review. 

df['sentiment'] = df['reviewText'].apply(get_sentiment)

df

reviewText  Positive  sentiment
0      one best apps acording bunch people agree bomb...         1          1
1      pretty good version game free . lot different ...         1          1
2      really cool game . bunch level find golden egg...         1          1
3      silly game frustrating , lot fun definitely re...         1          1
4      terrific game pad . hr fun . grandkids love . ...         1          1
...                                                  ...       ...        ...
19995  app fricken stupid.it froze kindle wont allow ...         0          0
19996  please add ! ! ! ! ! need neighbor ! ginger101...         1          1
19997  love ! game . awesome . wish free stuff house ...         1          1
19998  love love love app side fashion story fight wo...         1          1
19999  game rip . list thing make better & bull ; fir...         0          1

[20000 rows x 3 columns]

The NLTK sentiment analyzer returns a score between -1 and +1. We have used a cut-off threshold of 0 in the get_sentiment function above. Anything above 0 is classified as 1 (meaning positive).

Since we have actual labels, we can evaluate the performance of this method by building a confusion matrix. 

In [14]:
from sklearn.metrics import confusion_matrix

print(confusion_matrix(df['Positive'], df['sentiment']))

[[ 1131  3636]
 [  576 14657]]


The confusion_matrix function is used to evaluate the performance of a classification model by comparing the predicted labels with the actual labels.  
Then, the code prints the confusion matrix by passing two arguments to the confusion_matrix function: df['Positive'] and df['sentiment']. These two arguments are the actual labels and predicted labels, respectively.  
The rows of the matrix represent the actual labels, while the columns represent the predicted labels.   
The diagonal elements of the matrix represent the number of correct predictions, while the off-diagonal elements represent the number of incorrect predictions.

We can also check the classification report:

In [15]:
from sklearn.metrics import classification_report

print(classification_report(df['Positive'], df['sentiment']))

              precision    recall  f1-score   support

           0       0.66      0.24      0.35      4767
           1       0.80      0.96      0.87     15233

    accuracy                           0.79     20000
   macro avg       0.73      0.60      0.61     20000
weighted avg       0.77      0.79      0.75     20000



The overall accuracy of this rule-based sentiment analysis model is 79%.